pkg/virtctl/create/clone/clone.go

```diff
@@ -0,0 +1,286 @@
+
+	_ = cmd.MarkFlagRequired(SourceFlag)
+	_ = cmd.MarkFlagRequired(TargetFlag)
+
+	return cmd
+}
+
+func defaultSourceTarget(s *SourceTarget) {
+	if s.APIGroup == "" {
+		s.APIGroup = core.GroupName
+	}
+
+	if s.Kind == "" {
+		s.Kind = "VirtualMachine"
+	}
+}
+
+func validateSourceTarget(s SourceTarget, flag string) error {
+	if s.Name == "" {
+		return params.FlagErr(flag, fmt.Sprintf(MustBeSpecifiedErrFmt, "name"))
+	}
+	return nil
+}
+
+func sourceTargetToTypedLocalObjectReference(s SourceTarget) *v1.TypedLocalObjectReference {
+	return &v1.TypedLocalObjectReference{
+		APIGroup: pointer.P(s.APIGroup),
+		Kind:     s.Kind,
+		Name:     s.Name,
+	}
+}
+
+func withSource(c *createClone, cloneSpec *cloneSpec) error {
+	const flag = SourceFlag
+	source := SourceTarget{}
+
+	if !strings.ContainsAny(c.source, ",:") {
+		source.Name = c.source
+	} else {
+		err := params.Map(flag, c.source, &source)
+		if err != nil {
+			return err
+		}
+	}
+
+	defaultSourceTarget(&source)
+	err := validateSourceTarget(source, flag)
+	if err != nil {
+		return err
+	}
+
+	cloneSpec.Source = sourceTargetToTypedLocalObjectReference(source)
+	return nil
+}
+
+func withTarget(c *createClone, cloneSpec *cloneSpec) error {
+	const flag = TargetFlag
+	target := SourceTarget{}
+
+	if !strings.ContainsAny(c.target, ",:") {
+		target.Name = c.target
+	} else {
+		err := params.Map(flag, c.target, &target)
+		if err != nil {
+			return err
+		}
+	}
+
+	defaultSourceTarget(&target)
+	err := validateSourceTarget(target, flag)
+	if err != nil {
+		return err
+	}
+
+	cloneSpec.Target = sourceTargetToTypedLocalObjectReference(target)
+	return nil
+}
+
+func withNewMacAddresses(c *createClone, cloneSpec *cloneSpec) error {
+	const flag = NewMacAddressesFlag
+
+	for _, param := range c.newMacAddresses {
+		newMacAddresses := NewMacAddresses{}
+
+		err := params.Map(flag, param, &newMacAddresses)
+		if err != nil {
+			return err
+		}
+
+		if newMacAddresses.InterfaceName == "" {
+			return params.FlagErr(flag, fmt.Sprintf(MustBeSpecifiedErrFmt, "interfaceName"))
+		}
+		if newMacAddresses.NewMacAddress == "" {
+			return params.FlagErr(flag, fmt.Sprintf(MustBeSpecifiedErrFmt, "newMacAddress"))
+		}
+
+		if cloneSpec.NewMacAddresses == nil {
+			cloneSpec.NewMacAddresses = map[string]string{}
+		}
+
+		cloneSpec.NewMacAddresses[newMacAddresses.InterfaceName] = newMacAddresses.NewMacAddress
+	}
+
+	return nil
+}
+
+func (c *createClone) usage() string {
+	return `  # Create a manifest for a clone with a random name:
+  {{ProgramName}} create clone --source sourceVM --target targetVM
+  
+  # Create a manifest for a clone with a specified name:
+  {{ProgramName}} create clone --name my-clone --source sourceVM --target targetVM
+
+  # The above is a shortcut to:
+  {{ProgramName}} create clone --name my-clone --source name:sourceVM,apiGroup:kubevirt.io,kind:VirtualMachine --target name:targetVM,apiGroup:kubevirt.io,kind:VirtualMachine
```

Comment:
Why do we need to specify the `apiGroup` and `kind:VirtualMachine`? Couldn't we simply specify an additional flag `--type` and there you map type to the k8s object? This version seems a bit lengthy and not very straightforward to me.

Reply:
IMO, it will be nice to understand from the help option which kind of k8s objects I can clone

Reply:
> Why do we need to specify the apiGroup and kind:VirtualMachine? Couldn't we simply specify an additional flag --type and there you map type to the k8s object? This version seems a bit lengthy and not very straightforward to me.

TBH I'm in a dilemma here :)

The rationale for the flags as they are now is that it's fully generic in the sense that when the clone API would expand and change in the future, for example by supporting more source and target types, `virtctl create clone` command wouldn't have to be changed. If we would go with something like `--type`, then every supported type would need to be added to the mapping once supported, which might lead to inconsistencies and for harder time when expanding the API.

In addition, there's a need to specify a type for both the source and the target, which might be of different types. For example, the source could be a vmSnapshot while the target is a VM. TBH, I'm not sure how this can be mapped by order ,e.g. `virtctl create clone --source sourceSnapshot --type snapshot --target targetVM --type vm`, how can we know that the first `--type` refers to the source in this case?

Instead, we can simply omit the `apiGroup` and map it out automatically, e.g. `virtctl create clone --source name:sourceSnapshot,type:snapshot --target name:targetVM,type:vm`, but this is not a lot nicer and suffers from the disadvantage I laid out above.

Last point is that at least I've made VMs as the default, so cloning VMs can be done much for elegantly (`virtctl create clone --source sourceVM --target targetVM`).

But on the other hand - you're right - it is lengthy as for now.
What do you think? I'd appreciate your feedback here.

Reply:
> IMO, it will be nice to understand from the help option which kind of k8s objects I can clone

Same concern as above. While I think this is very valuable, I'm a bit concerned about not having one single source of truth. Updating in many places once a new type is supported gets me a bit uncomfortable.

WDYT about adding a link to the relevant user-guide entry instead?

Reply:
> In addition, there's a need to specify a type for both the source and the target, which might be of different types. For example, the source could be a vmSnapshot while the target is a VM. TBH, I'm not sure how this can be mapped by order ,e.g. `virtctl create clone --source sourceSnapshot --type snapshot --target targetVM --type vm`, how can we know that the first `--type` refers to the source in this case?

For this, if we can avoid the lengthy option, I still think a `--type-source` and `--type-target` are still more elegant :). If nothing is specified, you can always assume `VirtualMachine`.

> The rationale for the flags as they are now is that it's fully generic in the sense that when the clone API would expand and change in the future, for example by supporting more source and target types, virtctl create clone command wouldn't have to be changed. If we would go with something like --type, then every supported type would need to be added to the mapping once supported, which might lead to inconsistencies and for harder time when expanding the API.

Ok, yes this is the real question. Do you expect many k8s object types to be snapshottable? 
Additionally, if we have control of the type, we could validate a certain field. At least, here you are adding a new mac address option. This flag is only valid if the target is a VM. I'm not sure if the target (at least) can be an arbitrary k8s object. Same thing for the smbios

Reply:
> > IMO, it will be nice to understand from the help option which kind of k8s objects I can clone
> 
> Same concern as above. While I think this is very valuable, I'm a bit concerned about not having one single source of truth. Updating in many places once a new type is supported gets me a bit uncomfortable.
> 
> WDYT about adding a link to the relevant user-guide entry instead?

Well, if you have it in the code, for me that's the most reliable source of the truth :). I'd avoid putting links, they get outdated quickly. Let's try to understand what we prefer if an implicit or explicit list of the clonable types :). 
I don't want necessarily block this PR, only understand the reasoning behind

Reply:
> Ok, yes this is the real question. Do you expect many k8s object types to be snapshottable?

What I currently have in mind is:
1) VMI should be supported as both source and target types
2) We were once talking about supporting PVC as the source types. The main idea was that a VM state could reside in the PVC in some format, along with the VM manifest and perhaps a configuration file. We can also utilize the Export API for that.

It's hard to know which more object would need to be supported in the future. For example, maybe a `vmpool` could be cloned too?

But you're right, maybe we're not talking about a lot of new supported types.

I think I'm convinced :)
Thank you for this discussion

Reply:
> It's hard to know which more object would need to be supported in the future. For example, maybe a vmpool could be cloned too?

Maybe but if we really realize that the list starts to grow too much and we want to give more flexibility, we can always add an arbitrary type with the logic you currently have, but I would avoid it if not necessary